File: hashtable/525/525.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        hash = {}
        sum = 0
        ans = 0
        for i in range(len(nums)):
            sum += -1 if nums[i] == 0 else 1
            if sum == 0:
                ans = i + 1
            if sum not in hash:
                hash[sum] = i
            else:
                ans = max(ans, i - hash[sum])
        return ans

File: hashtable/525/525.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var findMaxLength = function (nums) {
  let sumToIndex = new Map();
  let sum = 0;
  let ans = 0;
  for (let i = 0; i < nums.length; i++) {
    sum += nums[i] === 0 ? -1 : 1;
    if (sum === 0) {
      ans = i + 1;
    }
    if (!sumToIndex.has(sum)) {
      sumToIndex.set(sum, i);
    } else {
      ans = Math.max(ans, i - sumToIndex.get(sum));
    }
  }
  return ans;
};
